engine/pkg/externalresource/s3/s3_file_manager.go

```diff
@@ -0,0 +1,254 @@
+	"go.uber.org/zap"
+)
+
+const (
+	placeholderFileName = ".keep"
+)
+
+// FileManager manages resource files stored on s3.
+type FileManager struct {
+	bucketSelector BucketSelector
+	storageFactory ExternalStorageFactory
+	index          indexManager
+	prefix         string
+}
+
+// NewFileManager creates a new FileManager.
+func NewFileManager(
+	executorID model.ExecutorID,
+	bucketSelector BucketSelector,
+	factory ExternalStorageFactory,
+) *FileManager {
+	return &FileManager{
+		bucketSelector: bucketSelector,
+		storageFactory: factory,
+		index:          newIndexManager(executorID, bucketSelector, factory),
+	}
+}
+
+// CreateResource creates a new resource on s3.
+func (m *FileManager) CreateResource(
+	ctx context.Context, ident internal.ResourceIdent,
+) (internal.ResourceDescriptor, error) {
+	bucket, err := m.bucketSelector.GetBucket(ctx, ident.Scope())
+	if err != nil {
+		return nil, errors.Annotate(err, "FileManager: CreateResource")
+	}
+
+	desc := newResourceDescriptor(bucket, ident, m.storageFactory)
+	storage, err := desc.ExternalStorage(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := createPlaceholderFile(ctx, storage); err != nil {
+		return nil, err
+	}
+	return desc, nil
+}
+
+// GetPersistedResource returns the descriptor of a resource that has already
+// been marked as persisted.
+// Note that GetPersistedResource will work on any executor for any persisted resource.
+func (m *FileManager) GetPersistedResource(
+	ctx context.Context, ident internal.ResourceIdent,
+) (internal.ResourceDescriptor, error) {
+	persistedResourceSet, err := m.index.LoadPersistedFileSet(ctx, ident.Scope())
+	if err != nil {
+		if gerrors.Is(err, errIndexFileDoesNotExist) {
+			return nil, internal.ErrResourceFilesNotFound.GenWithStack(
+				&internal.ResourceFilesNotFoundError{
+					Ident:   ident,
+					Details: err.Error(),
+				})
+		}
+		return nil, errors.Annotate(err, "FileManager: GetPersistedResource")
+	}
+	if _, ok := persistedResourceSet[ident.Name]; !ok {
+		return nil, internal.ErrResourceFilesNotFound.GenWithStack(
+			&internal.ResourceFilesNotFoundError{
+				Ident: ident,
+			})
+	}
+
+	bucket, err := m.bucketSelector.GetBucket(ctx, ident.Scope())
+	if err != nil {
+		return nil, errors.Annotate(err, "FileManager: GetPersistedResource")
+	}
+
+	desc := newResourceDescriptor(bucket, ident, m.storageFactory)
+	storage, err := desc.ExternalStorage(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	ok, err := storage.FileExists(ctx, placeholderFileName)
+	if err != nil {
+		return nil, errors.Annotate(err, "check placeholder file")
+	}
+	if !ok {
+		return nil, internal.ErrResourceFilesNotFound.GenWithStack(
+			&internal.ResourceFilesNotFoundError{
+				Ident: ident,
+			})
+	}
+
+	return desc, nil
+}
+
+// RemoveTemporaryFiles removes all temporary resources (those that are not persisted).
+// It can only be used to clean up resources created by the local executor.
```

Comment:
Better to check executorID within this function.

Reply:
I think the caller can guarantee it. This requirement is here due to the risk of data race in s3.